default.nix

```diff
@@ -78,29 +83,50 @@ stdenv.mkDerivation rec {
            !elem (baseNameOf path) [".git" "result" "bin" "_build" "_build_ci"]) ./.;
 
   preConfigure = ''
+    patchShebangs tools/
     patchShebangs dev/tools/
+    export COQ_CONFIGURE_PREFIX=$out
+  '' + optionalString doInstallCheck ''
+    patchShebangs test-suite/
+  '' + optionalString buildDoc ''
+    patchShebangs doc/tools/
   '';
 
-  prefixKey = "-prefix ";
+  createFindlibDestdir = !shell;
 
-  buildFlags = [ "world" "byte" ] ++ optional buildDoc "doc-html";
+  configurePhase = ''
+    runHook preConfigure
+    runHook postConfigure
+  '';
 
-  installTargets =
-    [ "install" "install-byte" ] ++ optional buildDoc "install-doc-html";
+  makefile = "Makefile.dune";
 
-  createFindlibDestdir = !shell;
+  buildFlags = [ "voboot" ];
 
-  postInstall = "ln -s $out/lib/coq $OCAMLFIND_DESTDIR/coq";
+  postBuild = ''
+    for package in ${toString packages}
+    do
+      dune build $package.install
+    done
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    for package in ${toString packages}
+    do
+      dune install --prefix=$out $package
+    done
+    ln -s $out/lib/coq $OCAMLFIND_DESTDIR/coq
```

Comment:
Why is this still needed? Anything we could do so we get the right thing done?

Reply:
You need to distinguish (in `dune-install`) the path where OCaml libraries are installed from the path where Coq libraries are installed.

Reply:
I thought we had fixed this. Currently we do:
```
lib_root: [
  "_build/install/default/lib/coq/plugins/btauto/.Algebra.aux" {"coq/plugins/btauto/.Algebra.aux"}
  "_build/install/default/lib/coq/plugins/btauto/.Btauto.aux" {"coq/plugins/btauto/.Btauto.aux"}
  "_build/install/default/lib/coq/plugins/btauto/.Reflect.aux" {"coq/plugins/btauto/.Reflect.aux"}
  "_build/install/default/lib/coq/plugins/btauto/Algebra.glob" {"coq/plugins/btauto/Algebra.glob"}
  "_build/install/default/lib/coq/plugins/btauto/Algebra.vo" {"coq/plugins/btauto/Algebra.vo"}
```
how is this not the right place? OCaml libraries go in `lib` which can be mapped to `lib/ocaml` by distributions for example.

Reply:
IMHO is very important we get this right as soon Dune itself will generate this rules and would be a pain to have to change them. In fact, the Dune version will place this stuff in `lib_root/coq/stdlib/...` as the lib will be named stdlib.

Reply:
Can you please elaborate on how distributions can map these different lib directories to what they need?

Reply:
I am not sure what distros need so I cannot elaborate a lot. In this case, you can use the `--libdir` and `--prefix` parameter to relocate `lib`, so you mean that `lib_root` should be not affected by `--libdir`?

What should the right layout be on Nix?

Reply:
Of course `dune-install` can be extended to accommodate Nix's needs.

Reply:
The issue here is that there are two `lib` directories. One for Coq, one for OCaml. I’m OK using `--libdir` to set one of them. What shall I do for the other one?

Currently, the layout used in nixpkgs is as follows (for non-standard libraries):
 - For OCaml: `$prefix/lib/ocaml/$ocaml-version/site-lib/`
 - For Coq: `$prefix/lib/coq/$coq-version/user-contrib`

Reply:
Yup I think this layout is not supported by dune-install; `--libdir` will move both library paths. We ought to add a new option to dune-install so we can properly remap `lib_root` at least.

I'll open an issue.

Reply:
> * For Coq: `$prefix/lib/coq/$coq-version/user-contrib`

The layout I propose for 8.11 [most likely is]:

`$prefix/lib/coq/$coq-version/$lib_name/`

does that sound good to you?

Reply:
In coq dune there will be no library loaded by default, users of the stdlib will have to declare it in the deps.

Reply:
Does this means getting rid of `COQPATH`?

Reply:
Both should coexist I think, as for example to keep the NIX use case, but a single nix package could install many coq packages in its private `lib/coq/` dir.

Reply:
Note that there is something attractive on installing coq packages under just `lib`, and this is that plugins get to be more self-contained.

Reply:
I am lost on one detail tho, what's the value of `$OCAMLFIND_DESTDIR` ?

Reply:
`$prefix/lib/ocaml/$ocaml_version/site-lib/`

Reply:
So indeed shouldn't you just use `dune install --libdir="$OCAMLFIND_DESDIR" --prefix="$out"` instead of `ln` ?

Reply:
Do you mean running that in addition to the usual install? Are you suggesting to install Coq in full twice?

Reply:
Umm, this should install Coq only once, but in the proper place, right?

Reply:
I mean as the default install call; `lib` dir is already allowed to be relocated by distros, so here we have two things: OCaml libraries installed in the wrong place, and Coq libraries installed in the wrong place.

For 1) `--libdir` is standard and supported, for 2) we need to think / tweak `dune install` I think.

Reply:
What do you call Coq?

First, we install the binaries, the stdlib, the doc… under $prefix.

Then, this line puts the OCaml libraries (for plug-in writers) where they can be found by `ocamlfind`.

Reply:
In fact I guess that all dune-built packages in Nix should follow this pattern; and indeed they do:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/ocaml/dune.nix#L24

So why should be Coq different here?

Reply:
You are right. As dune stands, since there is only one `lib` directory concept, you have to choose between OCaml libraries at the wrong place or Coq theories at the wrong place.

An other possibility could be to fix Coq so that the location of the theories is configurable (at configure time). With such a possibility, we could put both Coq and OCaml libraries in a single `lib` directory, where they can be found by `ocamlfind` as well as by `coqc`.

Reply:
Isn't the Nix Coq package setting `COQPATH` anyways? I thought that in your setup the original Coq configure location didn't matter so much, right?

IMHO we should first install OCaml stuff properly [as done in dune.nix], and indeed we should fix Dune so Coq libraries can be installed properly, for now they will be in the wrong place but at least on Nix [and OPAM] things should be fine as they use `COQPATH`.

Note that Coq's configure already has libdir which will allow to do that, Dune supports calling configure before the build so you can do: `./configure -prefix $out -libdir $OCAMLFIND_DESTDIR -native-compiler no` and then build normally.

I'll fix this native compiler stuff by the way, it is the only real annoying difference we've got.

Reply:
Thanks for repeating that again. I thought I had tried but I was doing the quoting wrong…